### ISSUE TITLE
[web] DeferredLoadException after deploying a new version of app

### ISSUE THREAD

**Issue by slavap** [flutter/flutter#127459]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

This is the same issue as closed https://github.com/flutter/flutter/issues/121210
When deferred loading is used, Flutter generates multiple js chunks, e.g. main.dart.js, main.dart.js_1.part.js, main.dart.js_2.part.js, ...
If browser has the previous versions of these chunks in the cache, it will just use them without fetching new versions from a server.
This issue can be workarounded by different ways, but the best one is to fix how part.js are loaded from main.dart.js by replacing 
**().createScriptURL(l+k) to ().createScriptURL(l+k+"?v=N.N.NN")**
That change will enforce the browser to always load part.js files when a new version is deployed.
Currently possible workarounds:
1. Ask users to clear the browser cache manually when a new version of app is deployed - totally "ugly" and inconvenient.
2. Add cache control attributes on the hosting server, i.e. at least no-cache and etag, but that will force the browser to ask a server every time and get 304 responses in most of the cases - so slow down considering that usually there are hundreds of part.js files.
3. Currently after 'flutter build web' I'm using the following post-processing script to fix this caching problem, but actually Flutter has to have this functionality embedded:
```bash
#!/bin/bash

# Based on: https://github.com/doonfrs/flutter-build-web-cache-problem/blob/main/build-web.sh
# Run only ONCE after 'flutter build web' as: ./prepare.sh
# Expected folder structure:
#   build 
#     web
#   prepare.sh
#   pubspec.yaml

echo "Reading version from pubspec.yaml without + sign"
version=$(grep version: pubspec.yaml | sed 's/version: //g' | sed 's/+//g')
echo "$version"

echo
echo "Calculating MD5"
md5_0=($(md5sum build/web/main.dart.js))
echo "main.dart.js   $md5_0"

md5_1=($(md5sum build/web/flutter.js))
echo "flutter.js   $md5_1"

md5_2=($(md5sum build/web/index.html))
echo "index.html   $md5_2"
echo

echo "Patching version in js partial urls in main.dart.js"
sed -i 's/\.createScriptURL([[:graph:]]\++[[:space:]]*\+[[:graph:]]/&+"?v='"$version"'"/g' build/web/main.dart.js

echo "Patching main.dart.js path in flutter.js"
sed -i 's|\(${baseUri}main\.dart\.js\)|\1?v='"$version"'|g' build/web/flutter.js

echo "Patching main.dart.js path in flutter.js (for obsolete Flutter 3.3)"
sed -i 's|\(= "main\.dart\.js\)|\1?v='"$version"'|g' build/web/flutter.js

echo "Patching index.html for flutter.js"
sed -i 's/"flutter\.js"/"flutter.js?v='"$version"'"/' build/web/index.html

echo
echo "Calculating new MD5"
md5_n0=($(md5sum build/web/main.dart.js))
echo "main.dart.js   $md5_n0"

md5_n1=($(md5sum build/web/flutter.js))
echo "flutter.js   $md5_n1"

md5_n2=($(md5sum build/web/index.html))
echo "index.html   $md5_n2"
echo

echo "Patching MD5 in flutter_service_worker.js"
sed -i 's/'"$md5_0"'/'"$md5_n0"'/' build/web/flutter_service_worker.js
sed -i 's/'"$md5_1"'/'"$md5_n1"'/' build/web/flutter_service_worker.js
sed -i 's/'"$md5_2"'/'"$md5_n2"'/' build/web/flutter_service_worker.js
```

### Expected results

Version query parameter has to be added to all js files generated during compilation.

### Actual results

DeferredLoadException: 'Loading main.dart.js_1.part.js failed: the code with hash ...'

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.7.12, on Microsoft Windows [Version 10.0.19045.2965], locale en-US)
    • Flutter version 3.7.12 on channel stable at c:\Programs\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4d9e56e694 (5 weeks ago), 2023-04-17 21:47:46 -0400
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\slavap\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = c:\Programs\chrome-debug.bat

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code (version 1.78.2)
    • VS Code at C:\Users\slavap\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.64.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2965]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 113.0.5672.127
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 113.0.1774.35

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

**Comment by huycozy** [comment]
Hi @slavap 
Thanks for filing the issue. Can you please provide a minimal sample code along with `Steps by steps` to reproduce issue?

**Comment by adar2378** [comment]
I also have this issue with deferred loading after deployment the hashes mismatch and it fails to load

**Comment by slavap** [comment]
@huycozy 
Complete error log in my case is the following (but it will be different depending from particular application and changes done in new app version):
```
DeferredLoadException: 'Loading main.dart.js_179.part.js failed: the code with hash 'EwrIJD9qMCMNQswubaWUM3omWvE=' was not loaded.
 event log:
  - _loadHunk: main.dart.js_76.part.js
  - download: main.dart.js_76.part.js from https://swuat/app/main.dart.js_76.part.js
  - _loadHunk: main.dart.js_77.part.js
  - download: main.dart.js_77.part.js from https://swuat/app/main.dart.js_77.part.js
  - _loadHunk: main.dart.js_117.part.js
  - download: main.dart.js_117.part.js from https://swuat/app/main.dart.js_117.part.js
  - download success: main.dart.js_76.part.js
  - initialize: main.dart.js_76.part.js (08o6smPADiyp5UcyHGK09ZqlJCc=)
  - download success: main.dart.js_77.part.js
  - initialize: main.dart.js_77.part.js (qOL/KPq0YUZSuNO46iIZZHV4T8o=)
  - download success: main.dart.js_117.part.js
  - initialize: main.dart.js_117.part.js (jMEFht9Wg4J8dNQ6ve2pN6qH+sk=)
  - _loadHunk: main.dart.js_110.part.js
  - download: main.dart.js_110.part.js from https://swuat/app/main.dart.js_110.part.js
  - download success: main.dart.js_110.part.js
  - already initialized: main.dart.js_76.part.js (08o6smPADiyp5UcyHGK09ZqlJCc=)
  - already initialized: main.dart.js_77.part.js (qOL/KPq0YUZSuNO46iIZZHV4T8o=)
  - initialize: main.dart.js_110.part.js (vsiQ9tBDD51ivvlOHA1YVR+x3dE=)
  - _loadHunk: main.dart.js_11.part.js
  - download: main.dart.js_11.part.js from https://swuat/app/main.dart.js_11.part.js
  - _loadHunk: main.dart.js_63.part.js
  - download: main.dart.js_63.part.js from https://swuat/app/main.dart.js_63.part.js
  - _loadHunk: main.dart.js_155.part.js
  - download: main.dart.js_155.part.js from https://swuat/app/main.dart.js_155.part.js
  - _loadHunk: main.dart.js_43.part.js
  - download: main.dart.js_43.part.js from https://swuat/app/main.dart.js_43.part.js
  - _loadHunk: main.dart.js_20.part.js
  - download: main.dart.js_20.part.js from https://swuat/app/main.dart.js_20.part.js
  - _loadHunk: main.dart.js_9.part.js
  - download: main.dart.js_9.part.js from https://swuat/app/main.dart.js_9.part.js
  - _loadHunk: main.dart.js_29.part.js
  - download: main.dart.js_29.part.js from https://swuat/app/main.dart.js_29.part.js
  - _loadHunk: main.dart.js_13.part.js
  - download: main.dart.js_13.part.js from https://swuat/app/main.dart.js_13.part.js
  - _loadHunk: main.dart.js_36.part.js
  - download: main.dart.js_36.part.js from https://swuat/app/main.dart.js_36.part.js
  - _loadHunk: main.dart.js_2.part.js
  - download: main.dart.js_2.part.js from https://swuat/app/main.dart.js_2.part.js
  - _loadHunk: main.dart.js_157.part.js
  - download: main.dart.js_157.part.js from https://swuat/app/main.dart.js_157.part.js
  - _loadHunk: main.dart.js_154.part.js
  - download: main.dart.js_154.part.js from https://swuat/app/main.dart.js_154.part.js
  - _loadHunk: main.dart.js_156.part.js
  - download: main.dart.js_156.part.js from https://swuat/app/main.dart.js_156.part.js
  - _loadHunk: main.dart.js_153.part.js
  - download: main.dart.js_153.part.js from https://swuat/app/main.dart.js_153.part.js
  - _loadHunk: main.dart.js_179.part.js
  - download: main.dart.js_179.part.js from https://swuat/app/main.dart.js_179.part.js
  - _loadHunk: main.dart.js_178.part.js
  - download: main.dart.js_178.part.js from https://swuat/app/main.dart.js_178.part.js
  - _loadHunk: main.dart.js_165.part.js
  - download: main.dart.js_165.part.js from https://swuat/app/main.dart.js_165.part.js
  - _loadHunk: main.dart.js_176.part.js
  - download: main.dart.js_176.part.js from https://swuat/app/main.dart.js_176.part.js
  - _loadHunk: main.dart.js_169.part.js
  - download: main.dart.js_169.part.js from https://swuat/app/main.dart.js_169.part.js
  - _loadHunk: main.dart.js_177.part.js
  - download: main.dart.js_177.part.js from https://swuat/app/main.dart.js_177.part.js
  - _loadHunk: main.dart.js_166.part.js
  - download: main.dart.js_166.part.js from https://swuat/app/main.dart.js_166.part.js
  - _loadHunk: main.dart.js_167.part.js
  - download: main.dart.js_167.part.js from https://swuat/app/main.dart.js_167.part.js
  - _loadHunk: main.dart.js_162.part.js
  - download: main.dart.js_162.part.js from https://swuat/app/main.dart.js_162.part.js
  - _loadHunk: main.dart.js_174.part.js
  - download: main.dart.js_174.part.js from https://swuat/app/main.dart.js_174.part.js
  - _loadHunk: main.dart.js_46.part.js
  - download: main.dart.js_46.part.js from https://swuat/app/main.dart.js_46.part.js
  - _loadHunk: main.dart.js_171.part.js
  - download: main.dart.js_171.part.js from https://swuat/app/main.dart.js_171.part.js
  - _loadHunk: main.dart.js_184.part.js
main.dart.js:35538  - download: main.dart.js_184.part.js from https://swuat/app/main.dart.js_184.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_198.part.js
main.dart.js:35538  - download: main.dart.js_198.part.js from https://swuat/app/main.dart.js_198.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_158.part.js
main.dart.js:35538  - download: main.dart.js_158.part.js from https://swuat/app/main.dart.js_158.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_180.part.js
main.dart.js:35538  - download: main.dart.js_180.part.js from https://swuat/app/main.dart.js_180.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_206.part.js
main.dart.js:35538  - download: main.dart.js_206.part.js from https://swuat/app/main.dart.js_206.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_187.part.js
main.dart.js:35538  - download: main.dart.js_187.part.js from https://swuat/app/main.dart.js_187.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_197.part.js
main.dart.js:35538  - download: main.dart.js_197.part.js from https://swuat/app/main.dart.js_197.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_172.part.js
main.dart.js:35538  - download: main.dart.js_172.part.js from https://swuat/app/main.dart.js_172.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_189.part.js
main.dart.js:35538  - download: main.dart.js_189.part.js from https://swuat/app/main.dart.js_189.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_199.part.js
main.dart.js:35538  - download: main.dart.js_199.part.js from https://swuat/app/main.dart.js_199.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_182.part.js
main.dart.js:35538  - download: main.dart.js_182.part.js from https://swuat/app/main.dart.js_182.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_170.part.js
main.dart.js:35538  - download: main.dart.js_170.part.js from https://swuat/app/main.dart.js_170.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_168.part.js
main.dart.js:35538  - download: main.dart.js_168.part.js from https://swuat/app/main.dart.js_168.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_193.part.js
main.dart.js:35538  - download: main.dart.js_193.part.js from https://swuat/app/main.dart.js_193.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_192.part.js
main.dart.js:35538  - download: main.dart.js_192.part.js from https://swuat/app/main.dart.js_192.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_163.part.js
main.dart.js:35538  - download: main.dart.js_163.part.js from https://swuat/app/main.dart.js_163.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_196.part.js
main.dart.js:35538  - download: main.dart.js_196.part.js from https://swuat/app/main.dart.js_196.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_188.part.js
main.dart.js:35538  - download: main.dart.js_188.part.js from https://swuat/app/main.dart.js_188.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_203.part.js
main.dart.js:35538  - download: main.dart.js_203.part.js from https://swuat/app/main.dart.js_203.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_164.part.js
main.dart.js:35538  - download: main.dart.js_164.part.js from https://swuat/app/main.dart.js_164.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_209.part.js
main.dart.js:35538  - download: main.dart.js_209.part.js from https://swuat/app/main.dart.js_209.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_210.part.js
main.dart.js:35538  - download: main.dart.js_210.part.js from https://swuat/app/main.dart.js_210.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_161.part.js
main.dart.js:35538  - download: main.dart.js_161.part.js from https://swuat/app/main.dart.js_161.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_204.part.js
main.dart.js:35538  - download: main.dart.js_204.part.js from https://swuat/app/main.dart.js_204.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_195.part.js
main.dart.js:35538  - download: main.dart.js_195.part.js from https://swuat/app/main.dart.js_195.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_211.part.js
main.dart.js:35538  - download: main.dart.js_211.part.js from https://swuat/app/main.dart.js_211.part.js
main.dart.js:35538  - _loadHunk: main.dart.js_205.part.js
main.dart.js:35538  - download: main.dart.js_205.part.js from https://swuat/app/main.dart.js_205.part.js
main.dart.js:35538  - download success: main.dart.js_11.part.js
main.dart.js:35538  - initialize: main.dart.js_11.part.js (3vUGGpbkFcF7KedDiiy9ui/BeNc=)
main.dart.js:35538  - download success: main.dart.js_63.part.js
main.dart.js:35538  - initialize: main.dart.js_63.part.js (6HDMSwJ9fhgubLuvbtNf+OQBS+c=)
main.dart.js:35538  - download success: main.dart.js_155.part.js
main.dart.js:35538  - initialize: main.dart.js_155.part.js (AEuDyFveLA/LHscB6iEY2rH6bAM=)
main.dart.js:35538  - download success: main.dart.js_43.part.js
main.dart.js:35538  - initialize: main.dart.js_43.part.js (RFxZCz3o5mojsas9wi9bNtuOb1Y=)
main.dart.js:35538  - download success: main.dart.js_20.part.js
main.dart.js:35538  - initialize: main.dart.js_20.part.js (ftTLoVU8JWSEo9piIpVgHj8Xl8c=)
main.dart.js:35538  - download success: main.dart.js_9.part.js
main.dart.js:35538  - initialize: main.dart.js_9.part.js (rQTw2d6fougnTF1hpN2grtss9jE=)
main.dart.js:35538  - download success: main.dart.js_29.part.js
main.dart.js:35538  - initialize: main.dart.js_29.part.js (E+03FwUxCneuc99iL1P2CNofk6s=)
main.dart.js:35538  - download success: main.dart.js_13.part.js
main.dart.js:35538  - initialize: main.dart.js_13.part.js (W48K6qIuCZM/2x/lc8PNn8U+Jwc=)
main.dart.js:35538  - download success: main.dart.js_36.part.js
main.dart.js:35538  - initialize: main.dart.js_36.part.js (ClLngN5mvhN4I6GPblSI4FLYMJM=)
main.dart.js:35538  - download success: main.dart.js_2.part.js
main.dart.js:35538  - initialize: main.dart.js_2.part.js (4Qu2+Tpy5fWQHBqwHSHcl6VHz5g=)
main.dart.js:35538  - download success: main.dart.js_157.part.js
main.dart.js:35538  - initialize: main.dart.js_157.part.js (MVGfW4hR8GATXURDIUatEuVY2I8=)
main.dart.js:35538  - download success: main.dart.js_154.part.js
main.dart.js:35538  - initialize: main.dart.js_154.part.js (Ocp9hXlD3PGW+u1TXpQRneo3Alw=)
main.dart.js:35538  - download success: main.dart.js_179.part.js
main.dart.js:35538  - download success: main.dart.js_156.part.js
main.dart.js:35538  - initialize: main.dart.js_156.part.js (NTZYQZnNpzRaDZcwMnQsh9z7VHE=)
main.dart.js:35538  - download success: main.dart.js_178.part.js
main.dart.js:35538  - download success: main.dart.js_153.part.js
main.dart.js:35538  - initialize: main.dart.js_153.part.js (K3jTBQCd2TZqyrgcG3+FejesfIU=)
main.dart.js:35538  - missing hunk: main.dart.js_179.part.js (EwrIJD9qMCMNQswubaWUM3omWvE=)
```

**Regarding steps to reproduce and minimal sample code - unfortunately, there are no simple steps to reproduce that problem, it is quite "random" by its nature. But despite of that, it is quite clear why this problem occurs.**

In code there should be multiple parts, which activated by deferred loading, something like that:
```dart
import 'package:sw/ui/guest/guest_edit_page.dart' deferred as _i14;
import 'package:sw/ui/home/home_page.dart' deferred as _i15;
import 'package:sw/ui/invoice/invoice_page.dart' deferred as _i17;
import 'package:sw/ui/login/login_page.dart' deferred as _i20;
import 'package:sw/ui/login/registration/registration_page.dart' deferred as _i21;

// And before using any of such pages, they first should be loaded, like:
await _i21.loadLibrary();
_i21.RegistrationPage();

await _i20.loadLibrary();
_i20.LoginPage();
```
But that's not all, there has to be changes in every new deployed version, so part.js files are different, and cached version of part.js is used instead of new and changed one.

**Comment by huycozy** [comment]
Hi @slavap 
The issue should be actionable with clear steps to reproduce it. Please try to provide a minimal sample code along with steps by steps reproduction so we can address the issue properly. You can check [this](https://github.com/perclasson/flutter_code/tree/master/deferred_loading) as a base sample and craft it to represent your case. Thank you!

**Comment by slavap** [comment]
@huycozy 
>> The issue should be actionable with clear steps to reproduce it.
Are you serious? That way you may fix only trivial problems.
For this case it is impossible to provide minimal sample code, because of:
1. There has to be hosting server where the application is deployed.
2. There should be TWO different versions of the application: "OLD" one and "NEW" one, and new one must have some changes comparing with old one to guarantee that part.js files are different for both versions.

Just forward this issue to developers for god's sake, it is obvious where the problem should be, it is widespread practice for SPA applications to add version as query parameter to its js script to prevent browser caching problems.

**Comment by adar2378** [comment]
Something related to this 
https://github.com/flutter/flutter/issues/63500

**Comment by slavap** [comment]
@huycozy Sorry for my previous a little bit aggressive comment. But “the OP didn't respond (for some reason)” for obvious reason - there is no simple and clear way to reproduce this problem, though why this problem occurs is quire obvious.

**Comment by yjbanov** [comment]
Optimistically redirecting this to the Dart SDK team, as the fix seems to be on the dart2js side. As a prior art, GWT used to do similar code splitting, but it would content-hash the part files and use the hash in the file name, which naturally busts all the caches sitting between the web page and the JS files.

**Comment by slavap** [comment]
@yjbanov 
Thank you! GWT deferred loading with putting js parts to subfolder with unique name was actually the very right solution.

Also there is another post-processing script for solving deferred loading problem: https://github.com/flutter/flutter/issues/63500#issuecomment-1562410001
Though it solves only part.js problem and do NOTHING with index.html reference to main.dart.js

**Comment by mkustermann** [comment]
> Optimistically redirecting this to the Dart SDK team, as the fix seems to be on the dart2js side. 

/cc @sigmundch

**Comment by sigmundch** [comment]
Today, the compiler doesn't have support for this and our recommendation has been to manage this issue as a serving infrastructure configuration. Our internal customers typically do so by directly encoding versioning on their serving paths, rather than using URL parameters. For example, the `index.html` includes a full path to the `main.dart.js` using a URL of the form `http://site/path/with/version-12345-release-may-31-20023/main.dart.js`. Unlike url parameters, this technique can keep the .js file names intact and doesn't require postprocessing of the output of the tools. Besides preventing invalid caching, this technique also allows to serve both old and new versions side-by-side. This is used heavily by internal customers to do incremental rollouts, which they control by serving a different `index.html` to a fraction of their users.

I'm less familiar with the serving infrastructure at play here, though. @yjbanov - what are the current recommendations for how to serve production apps? Does flutter provide a serving infrastructure or do we have a set of recommendations and each developer is creating their custom serving infrastructure? Is there something preventing this approach from working on the current serving infrastructure?

**Comment by slavap** [comment]
@sigmundch this approach with different deployment url for each new version works, but absolutely inconvenient for users.

**Comment by sigmundch** [comment]
Note that this versioned approach doesn't mean using a different path to reach the app itself. Instead, the home page would be on a stable and known location `http://mysite.com/index.html`. It is only the html file contents where a different path is used to fectch the `main.dart.js`. In other words, my expectation here is that end users are not aware of these various deployment URLs entirely.

@slavap - does that address your concern, or is there something else I'm missing?

**Comment by slavap** [comment]
@sigmundch No, it does not address my concerns :-)
First - there is no reference to main.dart.js in index.html at all. Have you looked at the compiled output?
There is 
```html
<script src="flutter.js" defer></script>
```
in index.html, which could be probably moved together with all main.dart.*.js files to a separate folder with a unique version-based name, but I did not try that and am not sure it will properly work.

**Comment by sigmundch** [comment]
😊 thanks for clarifying. I'm less familiar with flutter's deployment setup itself, so forgive me if my suggestions are not as accurate. Hopefully someone more familiar can chime in on that front

I honestly think it is possible to make this work as a serving solution, either with the current setup, with some changes to `flutter.js`, or with some additional redirection/indirection on the server side.

From a quick look at [the docs][1], I gather that the default setup is to put a .html that loads flutter.js, which then indirectly load the main.dart.js code. It appears that flutter.js loads `main.dart.js`, but allows using a [base tag][2] as url prefix if the app is not served at the root of the domain.

While `base-href` was not intended for versioning, it could potentially be used for it. It seems today that it is intended to mean that everything, even the .html and the flutter.js file, live under the version directory. I'd suggest trying if instead we could make it such that the index.html continues to live at the root, but all the other files are deeper in a versioned path and see how far it takes you.

@yjbanov - I think we should pursue improving the serving solution to support this and write up a recommendation on the flutter-web docs. This may require some fixes to `flutter.js` to make things work properly. Could you rope in those who currently maintain it?

[1]: https://docs.flutter.dev/deployment/web#deploying-to-the-web
[2]: https://docs.flutter.dev/ui/navigation/url-strategies#hosting-a-flutter-app-at-a-non-root-location

**Comment by slavap** [comment]
@sigmundch @yjbanov 
I think "base-href was not intended for versioning" - should remain as it is now, that is not good idea to change it for "versioning".

>> I'd suggest trying if instead we could make it such that the index.html continues to live at the root, but all the other files are deeper in a versioned path and see how far it takes you.

This solution sounds right to me, I'll do some testing to see if that works. But I'm afraid there is some logic in flutter.js currently, which expects that all files are placed in the same folder.

**Comment by yjbanov** [comment]
@sigmundch We do not offer a serving solution. Flutter Web apps are made of static files that can be uploaded to a plain site hosting service like Firebase or Netlify. Moving compiled files under a versioned directory works for `flutter.js`, but it doesn't work for users that simply put `<script src="main.dart.js">` in their HTML file. So I'm still leaning towards a compiler-based solution. Or we could create a second `main.dart.js` that wraps a `VERSION/main.dart.js`? 🤷

**Comment by yjbanov** [comment]
Optimistically accepting this issue on the Flutter side, although there's no elegant solution in my head yet. We could deprecate `main.dart.js` and migrate everyone to `flutter.js` (which is something we recommend already, but not everyone, including g3, is using it).

There's also the general question of how dart2js output is meant to be used. Should it be easy to just plop it into a `<script>` tag? Or is a wrapper necessary to make dart2js output work?

**Comment by sigmundch** [comment]
The more I think about this problem space, the more I believe this shouldn't be in the compiler hands 😊

In the future, we'd like to invest in refactoring the deferred loading logic, so it can be configured externally and treated as a user-level library, rather than a compiler internal thing. That could be a middle ground that allows for customization without having to include custom serving solutions within the compiler. That for example would enable dynamic bundling of multiple .part files more easily, for instance.

Re `<script>` tags - yes, the output is meant to be placed directly on a script tag. Even if flutter web doesn't own a serving infrastructure, I think we should have a proposed recommendation in docs for how to handle versioning properly. I worry that using flutter.js adds an unnecessary round trip to the initial load, so my inclination is to recommend using something like `<script src="VERSION/main.dart.js"></script>`, and make sure that the framework is flexible to allow the script code to be loaded from any path.

**Comment by slavap** [comment]
@yjbanov @sigmundch 
I did testing by moving everything to sub-folder (named as value taken from .last_build_id), like that:

![image](https://github.com/flutter/flutter/assets/4129763/740248bd-c28c-42f2-b05b-5c025f589759)

Obviously, that does not work, because flutter.js has "hardcoded" expectations about how files are placed and related to baseUri parameter as well. Also, there is a mention of index.html in flutter_service_worker.js, which should be changed somehow.

>> and migrate everyone to flutter.js

I don't see any problem with that, no reason to include main.dart.js directly in index.html

**So for now, post-processing scripts are the only working solution.** My script in the top of this issue and another one mentioned in https://github.com/flutter/flutter/issues/127459#issuecomment-1563637531

**Comment by dmitry-fbm** [comment]
After some back and forth we _seems_ (more testing to be done) were able to solve the issue by disabling caching of `index.html` along with removal of `--source-maps --no-frequency-based-minification` build options. However this solution is far from ideal because it also disables caching of `main.dart.js`. For instance, sending a user for 3DS authentication and then back causes unnecessary complete reload of 1.5M file.

**Comment by wegn8208** [comment]
ran into same situation here, 
I was using flutter web to build my website with firebase hosting.
One day I feel like the `main.dart.js` is way too big(3.1MB), so I used `deferred loading` to downsize it ( it is 2MB now, and could be better, and that is not counting hosting compress)

The first deploy was nice, but the later deployments are all broken.... 
It is not a random problem if someone is building a production web like me, I have more than 100 `main.dart.part_XXX.js`, every little change makes the part file a little different. for sure it will broke something.

--
my work around here is 
- build with build number and name, triggers service worker js file cache invalidate:
    ` flutter build web --build-name=VERSION_NAME --build-number=VERSION_NUMBER`
- manually modify flutter.js  in index.html with versioning
      ` <script src="flutter.js?vVERSION_NAME" defer></script>`
- find *_loadEntrypoint* in *flutter.js*, and change the line close to it, this should solve the *main.dart.js* cache problem
`   const scriptTag = this._createScriptTag(entrypointUrl + "?vVERSION_NAME");`
- find `dartDeferredLibraryLoader` in *main.dart.js* and change the following line, it should solve the *\*part.js* cache problem
`e.src=j + "?vVERSION_NAME"`

**Comment by mminhlequang** [comment]
Same
I got this if use `deferred` and manage version main.dart.js
example: from `/main.dart.js => `/main.dart.js?v=1.0.39`

**Comment by mminhlequang** [comment]
up

**Comment by JeffOlajos** [comment]
We have the same issue. As @wegn8208 stated,  

dartDeferredLibraryLoader in main.dart.js and change the following line, it should solve the *part.js cache problem
e.src=j + "?vVERSION_NAME"

wil fix the issue. Having to manually edit this file every build is problematic and seems like a simple mod to just generate the main.daryt.js to do this automatically for the user.

**Comment by slavap** [comment]
@JeffOlajos 
it is not only one main.dart.js file - there are at least 4 files to be changed, see the patch script.

**Comment by JeffOlajos** [comment]
We have our own custom index.html that takes care of the other issues. With
just the single change everything works as it should. I can send it to you
if you like or you can find it on our Flutter Markup Language GIT repo.
You can find a link to that on our site. Https://fml.dev.

Thanks !

**Comment by JeffOlajos** [comment]
The other item that’s worth noting is you also need to prevent index.html
from caching as well. You need to add meta tags to do that. We ran into the
issues where we were able to prevent flutter.js, main.dart.js and all the
deferred parts from caching but if the browser caches index.html that’s all
for not.

**Comment by slavap** [comment]
@JeffOlajos 
meta tags are not preventing index.html from caching - to properly resolve this problem you have to configure your hosting server properly to return no-cache in response, no other way to fix it as I know.

**Comment by mminhlequang** [comment]
any tmp solution for this issue?

**Comment by slavap** [comment]
@mminhlequang 
for now, post-processing scripts are the only working solution. My script in the top of this issue and another one mentioned in https://github.com/flutter/flutter/issues/127459#issuecomment-1563637531

**Comment by JeffOlajos** [comment]
We use a simple fix that only requires changing one line in main.dart.js.
We’ve written a custom index.html that can be found in our gut repo. It’s
referenced here: http://fml.dev.  You can modify index.html title as
required and remove script includes.  The line in main.dart.js that needs
to change is in the comments inside index.html.  Any questions just shoot
me an email.

**Comment by DamienMrtl** [comment]
We are now in 2024, I think we should get some traction and fix this issue.

**Comment by reza00farjam** [comment]
Any working temporary solution or workaround for new [flutter web app initialization](https://docs.flutter.dev/platform-integration/web/initialization)?

**Comment by slavap** [comment]
@reza00farjam updated bash script to support Flutter 3.24

**Comment by reza00farjam** [comment]
@slavap Thanks for reaching out. Couldn't find the updated script you mentioned. Can you provide a link to it? Thanks in advance.

**Comment by slavap** [comment]
@reza00farjam First message :-)
https://github.com/flutter/flutter/issues/127459#issue-1723105251

**Comment by slavap** [comment]
@yjbanov @sigmundch @reza00farjam 
By the way, I've decided to try again the solution with placing compiled app to subfolder named as app version.
And to my absolute surprise - looks like everything is working with Flutter 3.24 (not sure about earlier versions).
And post-processing script is not needed anymore!
What exactly I did:
1. flutter build web --release
2. Then copied build/web folder to my webserver and renamed it as web1234.
3. Moved index.html out of web1234 folder, so now index.html and web1234 folder are on the same level.
    Also, for better PWA support, the original index.html may remain in web1234 folder (no need to edit it), modify only the copied one.
4. Edited index.html as:
```html
        <head>
          <base href="web1234/">
```

5. Looks like everything is working now (needs additional testing) and no problem with "stuck" caching in the browser, because next deployments will go to the different folders, like web1235, web1236, ...

**Comment by uldall** [comment]
> 4\. dited index.html as:
> 
> 
> ```
>         <head>
>           <base href="web1234/">
> ```

@slavap You can leave the editing of index.html out by doing this:
```
  <!--
    If you are serving your web app in a path other than the root, change the
    href value below to reflect the base path you are serving from.

    The path provided below has to start and end with a slash "/" in order for
    it to work correctly.

    For more details:
    * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

    This is a placeholder for base href that will be replaced by the value of
    the `--base-href` argument provided to `flutter build`.
  -->
  <base href="$FLUTTER_BASE_HREF">
```

**Comment by slavap** [comment]
@uldall 
>> The path provided below has to start and end with a slash "/" in order for it to work correctly.

Sounds scary :-) Actually, in my case, there is NO leading / in base href and it **shouldn't** be there. But it works just fine.
Also, I don't see any explicit mention of "has to start and end with a slash /" in the documentation, where is that requirement coming from?

**Comment by uldall** [comment]
@slavap The comment is written in the default `web/index.html` generated by `flutter create`. Yeah I am also not sure about that leading `/`. Just copied what was in the official documentation :-)

**Comment by katjai1** [comment]
Similar issue when we use use_deferred_loading for flutter_intl. 

```
DeferredLoadException: 'Loading https://m.somesite.com/main.dart.js_2.part.js?dart2jsRetry=3 failed: Success callback invoked but part main.dart.js_2.part.js not loaded.
Context: 
event log:
{"p":"main.dart.js_2.part.js","e":"downloadFailure","l":"messages_en","s":"https://m.datingawidower.net/main.dart.js","i":78}
{"p":"main.dart.js_2","e":"endPart","h":"51IKErIQufLTiRFFaVbZeQ8lS5k=","i":77}
{"p":"main.dart.js_2","e":"beginPart","i":76}
{"p":"main.dart.js_32.part.js","e":"download","l":"messages_en","i":75}
{"p":"main.dart.js_32.part.js","e":"startLoad","l":"messages_en","i":74}
{"p":"main.dart.js_32.part.js","e":"retry2","l":"messages_en","i":73}
{"p":"main.dart.js_32","e":"endPart","h":"Gj0gzJeWVDCfvPMptpQ9iautQXs=","i":72}
{"p":"main.dart.js_32","e":"beginPart","i":71}
{"p":"main.dart.js_2.part.js","e":"download","l":"messages_en","i":70}
{"p":"main.dart.js_2.part.js","e":"startLoad","l":"messages_en","i":69}
{"p":"main.dart.js_2.part.js","e":"retry2","l":"messages_en","i":68}
{"p":"main.dart.js_2","e":"endPart","h":"51IKErIQufLTiRFFaVbZeQ8lS5k=","i":67}
{"p":"main.dart.js_2","e":"beginPart","i":66}
{"p":"main.dart.js_32.part.js","e":"download","l":"messages_en","i":65}
{"p":"main.dart.js_32.part.js","e":"startLoad","l":"messages_en","i":64}
{"p":"main.dart.js_32.part.js","e":"retry1","l":"messages_en","i":63}
{"p":"main.dart.js_32","e":"endPart","h":"Gj0gzJeWVDCfvPMptpQ9iautQXs=","i":62}
{"p":"main.dart.js_32","e":"beginPart","i":61}
{"p":"main.dart.js_2.part.js","e":"download","l":"messages_en","i":60}
{"p":"main.dart.js_2.part.js","e":"startLoad","l":"messages_en","i":59}
{"p":"main.dart.js_2.part.js","e":"retry1","l":"messages_en","i":58}
{"p":"main.dart.js_2","e":"endPart","h":"51IKErIQufLTiRFFaVbZeQ8lS5k=","i":57}
{"p":"main.dart.js_2","e":"beginPart","i":56}
{"p":"main.dart.js_32.part.js","e":"download","l":"messages_en","i":55}
{"p":"main.dart.js_32.part.js","e":"startLoad","l":"messages_en","i":54}
{"p":"main.dart.js_32.part.js","e":"retry0","l":"messages_en","i":53}
{"p":"main.dart.js_32","e":"endPart","h":"Gj0gzJeWVDCfvPMptpQ9iautQXs=","i":52}
{"p":"main.dart.js_32","e":"beginPart","i":51}
{"p":"main.dart.js_2.part.js","e":"download","l":"messages_en","i":50}
{"p":"main.dart.js_2.part.js","e":"startLoad","l":"messages_en","i":49}
{"p":"main.dart.js_2.part.js","e":"retry0","l":"messages_en","i":48}
{"p":"main.dart.js_2","e":"endPart","h":"51IKErIQufLTiRFFaVbZeQ8lS5k=","i":47}
{"p":"main.dart.js_2","e":"beginPart","i":46}
{"p":"main.dart.js_32.part.js","e":"download","l":"messages_en","i":45}
{"p":"main.dart.js_32.part.js","e":"startLoad","l":"messages_en","i":44}
{"p":"main.dart.js_2.part.js","e":"download","l":"messages_en","i":43}
{"p":"main.dart.js_2.part.js","e":"startLoad","l":"messages_en","i":42}
{"p":"main.dart.js_32.part.js","e":"downloadFailure","l":"messages_en","i":41}
{"p":"main.dart.js_32","e":"endPart","h":"Gj0gzJeWVDCfvPMptpQ9iautQXs=","i":40}
{"p":"main.dart.js_32","e":"beginPart","i":39}
...
```

**Comment by slavap** [comment]
@katjai1 did you tried to post-process your js after build with bash script from the first message?
OR base href workaround few messages above?

**Comment by katjai1** [comment]
> @katjai1 did you try to post-process your js after build with bash script from the first message? OR base href workaround few messages above?

We have a bit more tricky post-processing stage, because single codebase is used for a lot of sites. In result we build own  flutter_service_worker_[VERSION].js.base file, which is additionally post-processed using special web server script so that add to there  site-specific resources. But in general an idea is almost the same. GWT way with special directory looks much better. I actually think about base href way right now.

**Comment by slavap** [comment]
@tainanfochesatto 
>> path URL strategy to remove the hash (#) from the URL

Luckily it's used not so often, because requires additional support from the server (if I'm not mistaken).

Anyway, it would be nice to have this problem resolved, still not clear to me where and why "the subfolder (base href) is appended to the URL" ?

**Comment by philitell** [comment]
> > 4. dited index.html as:
> > ```
> >         <head>
> >           <base href="web1234/">
> > ```
> 
> @slavap You can skip the editing of index.html by doing this:
> 
> ```
>   
>   <base href="$FLUTTER_BASE_HREF">
> ```

I can confirm that you can skip editing the index.html-file by using the "--base-href" argument for flutter build. This way you can also automate the deployment process without any special in-file operations. 
E.g. when i'm deploying version 1.1.2:
```
flutter build web --base-href "/my_project_1_1_2/"
```
I'm using the base-href argument for build and a simple bash-deploy script that moves the index.html file to the parrent folder on the server. So the final file/folder-structure looks like:
```
my_project/
   index.html
   my_project_1_1_2/
```
For deployment of a new version i'm deleting all files and folders in "my_project" folder and perform a deployment as before with a new subfolder name e.g. "my_project_1_1_3".
AAAAND all problems with caching after version-update in a flutter-web-project are gone!!
Thank you @slavap for this simple workaround!!

**Comment by larsbloch** [comment]
@philitell 
But this makes changes to the url

![image](https://github.com/user-attachments/assets/07f50d26-ddbd-441f-bc43-12f28ea455bc)

Is there a way to avoid that ?

**Comment by larsbloch** [comment]
@slavap 
Thanks for getting back.

I get another URL
`
Old url
https://domain.me/subpage/subpage

New url
https://domain.me/buildNo_115/subpage/subpage`


When adding all the files to the subfolder i get the subfolder in my url of my project

Im, building like this
`RUN flutter build web --release --base-href "/buildNo_${BUILDNO}/"`

My indexfile looks like this
```
 <head>
  <base href="$FLUTTER_BASE_HREF">
```

So my files are like this

```
my_project/

   index.html

   buildNo_115/
```

I think NGINX is changing the url. Maybe its a me problem in the way im serving my application

**Comment by philitell** [comment]
@larsbloch I'm not using flutter's url path strategy and nginx on my server - there is no change in the url (in terms of the urls that are used (and served) by the flutter application). Of course the url of the assets change (absolut path of image changes - but the app handles this because of the base-href Setting) - but thats the point of this strategy.

**Comment by larsbloch** [comment]
I think i got something to work. I have tested locally and on our dev environment with a new build. I will post an update when i have done some more testing and i have changes.

My setup is maybe different as i build a docker image using nginx. I thought the solution mentioned before would be universal. But nevertheless it gave me inspiration to find a solution.

I have spent some time reading (and conversing with chatgpt) up on how webfiles are served (im a backender).
Changind the directory on nginx is enough in my case

That means this is in my nginx config
```
server {
    location / {
        root /usr/share/nginx/html/buildNo_${BUILDNO};
        index index.html index.htm;
        try_files $uri $uri/ /index.html =404;
    }
}
```

And then i have a github build parameter set into my dockerfile and added as a foldername
```
COPY --from=build-env /app/build/web /usr/share/nginx/html/buildNo_${BUILDNO}
COPY --from=build-env /app/.dockerResources/nginx.conf /etc/nginx/conf.d/default.conf
RUN sed -i "s/\${BUILDNO}/${BUILDNO}/g" /etc/nginx/conf.d/default.conf
```

I think this is a clean solution. I hope there isnt anything i have missed.

**Comment by philitell** [comment]
@slavap Up to this point i've done multiple releases using the subfolder-strategy:
 - https://github.com/flutter/flutter/issues/127459#issuecomment-2378870433
 - exactly as i have explained here: https://github.com/flutter/website/pull/11397#issuecomment-2520511842

I have to say that it does NOT work reliably. Today i've made an deployment and i was able to use the old cached version of my web-app until a hard refresh (CTRL+F5). The subfolder-strategy had no effect in this particular case (i've set all version numbers, changed directory name etc). On other occasions it worked fine but i didn't manage to find out why it did. A wild speculation: it seems not to work if you have a web-page in your local browser cache for a very long time and haven't visited the page for a long time.
The web-page is hosted on an ubuntu-server with nginx.

**Comment by slavap** [comment]
@philitell 
With subfolder strategy the only problem I can imagine - is cached index.html by browser, but for it you have to configure your hosting to send strict no-cache.

**Comment by philitell** [comment]
I configured cache-behaviour for index.html and version.json - and now it works. It seems subfolder-strategy + cache settings for this two files is the way to go.

**Comment by slavap** [comment]
@slavap version.json remains in subfolder, so no need to tune no-cache for it, only for index.html

**Comment by harkairt** [comment]
> [@philitell](https://github.com/philitell) But this makes changes to the url
> 
> ![image](https://github.com/user-attachments/assets/07f50d26-ddbd-441f-bc43-12f28ea455bc)
> 
> Is there a way to avoid that ?

I do have the same issue. Although I only tried to reproduce locally, but regardless what I try it with. `http-serve`, `serve` or `ngnix` the base href is being put into the url. 
Regardless of my `<base href="..">` is starting with a leading `/` or not (although according to the comment which got generated into the html it should have a leading /
>     The path provided below has to start and end with a slash "/" in order for it to work correctly.
but maybe this is an outdated comment)

**Comment by reza00farjam** [comment]
Great workaround @slavap ! Just a heads-up: don’t forget to change the `start_url` in `manifest.json` from the default `.` to `..`. Without this change, the PWA added to the home screen (especially on iOS Safari) will try to launch from `https://domain.com/<version>/`, which can lead to errors.

**Comment by slavap** [comment]
@reza00farjam what if the original unchanged index.html also remains in version123 folder? Then perhaps no need to edit manifest.json

**Comment by adar2378** [comment]
Guys, if anyone using firebase hosting, you can try --force flag to update all the files and invalidate old ones.

**Comment by DanilKarasev** [comment]
> I configured cache-behaviour for index.html and version.json - and now it works. It seems subfolder-strategy + cache settings for this two files is the way to go.

Hey, I'm trying to convince the team to use this approach, but they are saying index.html is still gonna be cached even if you set cache-control headers to `no-cache` unless you using `--pwa-stategy none` flag.

Are you building web app with this flag?

**Comment by alshareeforgect-pixel** [comment]
This one looks good outside and it doesn't even unfortunately get the full spectrum on

**Comment by slavap** [comment]
@DanilKarasev 
>> but they are saying index.html is still gonna be cached even if you set cache-control headers to no-cache

Browser won't cache index.html if no-cache is sent in response by the server, you can even respond with more strict "no-cache, must-revalidate, max-age=0".

**Comment by xiqishow** [comment]
After investigation, the issue appears to be caused by browser caching: the generated part.js file gets cached. To address this, I added a version query parameter when loading the JS file (e.g., ?v=123456). This approach preserves the caching mechanism while ensuring that when a new version is released and the version number changes, the browser will automatically fetch the latest file.
I'm using Flutter 3.27.3. To resolve the caching issue, I modified the build output as follows. ${hver} is the version number generated by Jenkins.The version is injected into the JS URLs during the post-build step.
`sed -i '' "s/i\.src=g/i.src=g + '?v=${hver}'/g" main.dart.js`
`sed -i '' "s|\"mainJsPath\":\"main.dart.js\"|\"mainJsPath\":\"main.dart.js?v=${hver}\"|g" flutter_bootstrap.js`
`sed -i '' -E "s|flutter_bootstrap\.js|flutter_bootstrap.js?v=${hver}|g" index.html`
Additionally, I've integrated the code from #63500. It's running smoothly so far, and I hope this serves as a helpful reference.
